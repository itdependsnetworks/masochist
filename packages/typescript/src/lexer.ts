import {ignore, union} from '@masochist/lexer';

// Operators.
export const ASSIGN = '=';
export const BANG = '!';
export const CLOSING_BRACE = '}';
export const CLOSING_BRACKET = ']';
export const CLOSING_PAREN = ')';
export const COMMA = ',';
export const COLON = ':';
export const DOT = '.';
export const EQUALS = '==';
export const GREATER_THAN = '>';
export const GREATER_THAN_OR_EQUAL = '>=';
export const LESS_THAN = '<';
export const LESS_THAN_OR_EQUAL = '<=';
export const LOGICAL_AND = '&&';
export const LOGICAL_OR = '||';
export const MINUS = '-';
export const OPENING_BRACE = '{';
export const OPENING_BRACKET = '[';
export const OPENING_PAREN = '(';
export const PLUS = '+';
export const REST = '...';
export const SEMICOLON = ';';
export const STRICT_EQUALS = '===';

// Reserved words.
//
// See: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Lexical_grammar#reserved_words
export const CLASS = 'class';
export const CONST = 'const';
export const CONTINUE = 'continue';
export const DEFAULT = 'default';
export const ELSE = 'else';
export const EXPORT = 'export';
export const FALSE = 'false';
export const FOR = 'for';
export const FUNCTION = 'function';
export const IF = 'if';
export const IMPORT = 'import';
export const LET = 'let';
export const NEW = 'new';
export const NULL = 'null';
export const RETURN = 'return';
export const THIS = 'this';
export const TRUE = 'true';
export const WHILE = 'while';
export const YIELD = 'yield';

// Special identifiers.
// See: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Lexical_grammar#identifiers_with_special_meanings

export const GET = 'get';
// others: arguments, as, async, eval, from, get, of, set

// TODO: add $ here
// don't bother with non-ASCII identifiers as described in:
// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Lexical_grammar#identifiers
export const IDENTIFIER = /[_a-z][_0-9a-z]*/i;
export const NUMBER = /-?(0|[1-9]\d*)(\.\d+)?/;

// String literals.
// See: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Grammar_and_types

const OCTAL_ESCAPE = /\\[0-7][0-7][0-7]/;
const HEX_ESCAPE = /\\x[0-9A-Fa-f][0-9A-Fa-f]/;
const UNICODE_ESCAPE = /\\u[0-9A-Fa-f][0-9A-Fa-f][0-9A-Fa-f][0-9A-Fa-f]/;
const EXTENDED_UNICODE_ESCAPE =
  /\\u\{[0-9A-Fa-f][0-9A-Fa-f][0-9A-Fa-f][0-9A-Fa-f][0-9A-Fa-f]\}/;
const CHARACTER_ESCAPE = /\\./;

const SINGLE_QUOTED_STRING = "'(" +
  [
    OCTAL_ESCAPE,
    HEX_ESCAPE,
    UNICODE_ESCAPE,
    EXTENDED_UNICODE_ESCAPE,
    CHARACTER_ESCAPE,
    /[^'\\]/,
  ]
    .map((regExp) => regExp.toString().slice(1, -1))
    .join('|') +
  ")*'";
const DOUBLE_QUOTED_STRING = '"(' +
  [
    OCTAL_ESCAPE,
    HEX_ESCAPE,
    UNICODE_ESCAPE,
    EXTENDED_UNICODE_ESCAPE,
    CHARACTER_ESCAPE,
    /[^"\\]/,
  ]
    .map((regExp) => regExp.toString().slice(1, -1))
    .join('|') +
  ')*"';

export const STRING_VALUE = new RegExp(
  SINGLE_QUOTED_STRING + '|' + DOUBLE_QUOTED_STRING,
);

// Ignored.
export const COMMENT = /\/\/[^\n\r]*(?:\r?\n)?/;
export const LINE_TERMINATOR = /\n|\r\n|\r/; // Note: we require semicolons.
export const WHITESPACE = /[\t ]+/;

export default union({
  // Order matters; reserved keywords and special identifiers must come before
  // normal identifiers.
  CLASS,
  CONST,
  CONTINUE,
  DEFAULT,
  ELSE,
  EXPORT,
  FALSE,
  FOR,
  FUNCTION,
  GET,
  IF,
  IMPORT,
  LET,
  NEW,
  NULL,
  RETURN,
  THIS,
  TRUE,
  WHILE,
  YIELD,

  ASSIGN,
  BANG,
  CLOSING_BRACE,
  CLOSING_BRACKET,
  CLOSING_PAREN,
  COLON,
  COMMA,
  DOT,
  EQUALS,
  GREATER_THAN,
  GREATER_THAN_OR_EQUAL,
  IDENTIFIER,
  LESS_THAN,
  LESS_THAN_OR_EQUAL,
  LOGICAL_AND,
  LOGICAL_OR,
  MINUS,
  NUMBER,
  OPENING_BRACE,
  OPENING_BRACKET,
  OPENING_PAREN,
  PLUS,
  REST,
  SEMICOLON,
  STRICT_EQUALS,
  STRING_VALUE,

  IGNORED: ignore(COMMENT, LINE_TERMINATOR, WHITESPACE),
});
