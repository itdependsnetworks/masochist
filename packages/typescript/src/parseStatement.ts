// @ts-nocheck
/**
 * vim: set nomodifiable : DO NOT EDIT - edit "build.ts", run "make graphql" instead
 *
 * @generated
 */
import {Lexer, Token} from './lex';
/**
 * r0: no production
 */
function r1($1) {
  return $1;
}
function r2($1) {
  return $1;
}
function r3($1) {
  return $1;
}
function r4(_$1, $2, _$3, $4) {
  return {
    kind: 'AssignmentStatement',
    binding: 'const',
    lhs: $2,
    rhs: $4,
  };
}
function r5(_$1, $2, _$3, $4) {
  return {
    kind: 'AssignmentStatement',
    binding: 'let',
    lhs: $2,
    rhs: $4,
  };
}
function r6($1) {
  return {
    kind: 'Identifier',
    name: $1.contents,
  };
}
function r7($1) {
  return $1;
}
function r8($1) {
  return $1;
}
function r9($1) {
  return $1;
}
function r10($1) {
  return $1;
}
function r11($1) {
  return $1;
}
function r12($1) {
  return $1;
}
function r13() {
  return {
    kind: 'ArrayValue',
    items: [],
  };
}
function r14(_$1, $2) {
  return {
    kind: 'ArrayValue',
    items: $2,
  };
}
function r15($1) {
  return [$1];
}
function r16($1, _$2, $3) {
  $1.push($3);
  return $1;
}
function r17() {
  return {
    kind: 'BooleanValue',
    value: false,
  };
}
function r18() {
  return {
    kind: 'BooleanValue',
    value: true,
  };
}
function r19() {
  return {
    kind: 'NullValue',
  };
}
function r20($1) {
  return {
    kind: 'NumberValue',
    value: parseFloat($1.contents),
    base: 10,
  };
}
function r21(_$1, _$2, $3) {
  return {
    kind: 'ExportDefaultDeclaration',
    declaration: $3,
  };
}
function r22($1) {
  return $1;
}
function r23(_$1, $2) {
  return {
    kind: 'ClassDeclaration',
    id: $2.name,
    body: [],
  };
}
function r24(_$1, $2, _$3, $4) {
  return {
    kind: 'ClassDeclaration',
    id: $2.name,
    body: $4,
  };
}
function r25($1) {
  return [$1];
}
function r26($1, $2) {
  $1.push($2);
  return $1;
}
function r27($1) {
  return $1;
}
function r28($1, _$2, $3) {
  return {
    kind: 'PropertyDeclaration',
    name: $1.name,
    type: $3.name,
  };
}
const actions = [
  {
    CLASS: {
      kind: 'Shift',
      state: 6,
    },
    CONST: {
      kind: 'Shift',
      state: 3,
    },
    EXPORT: {
      kind: 'Shift',
      state: 8,
    },
    LET: {
      kind: 'Shift',
      state: 4,
    },
  },
  {
    ['$']: {
      kind: 'Accept',
    },
  },
  {
    ['$']: {
      kind: 'Reduce',
      rule: 1,
    },
  },
  {
    IDENTIFIER: {
      kind: 'Shift',
      state: 10,
    },
  },
  {
    IDENTIFIER: {
      kind: 'Shift',
      state: 10,
    },
  },
  {
    ['$']: {
      kind: 'Reduce',
      rule: 2,
    },
  },
  {
    IDENTIFIER: {
      kind: 'Shift',
      state: 10,
    },
  },
  {
    ['$']: {
      kind: 'Reduce',
      rule: 3,
    },
  },
  {
    DEFAULT: {
      kind: 'Shift',
      state: 13,
    },
  },
  {
    ASSIGN: {
      kind: 'Shift',
      state: 14,
    },
  },
  {
    ASSIGN: {
      kind: 'Reduce',
      rule: 6,
    },
    OPENING_BRACE: {
      kind: 'Reduce',
      rule: 6,
    },
    SEMICOLON: {
      kind: 'Reduce',
      rule: 6,
    },
    COLON: {
      kind: 'Reduce',
      rule: 6,
    },
    CLOSING_BRACKET: {
      kind: 'Reduce',
      rule: 6,
    },
    COMMA: {
      kind: 'Reduce',
      rule: 6,
    },
  },
  {
    ASSIGN: {
      kind: 'Shift',
      state: 15,
    },
  },
  {
    OPENING_BRACE: {
      kind: 'Shift',
      state: 16,
    },
  },
  {
    CLASS: {
      kind: 'Shift',
      state: 6,
    },
  },
  {
    FALSE: {
      kind: 'Shift',
      state: 25,
    },
    IDENTIFIER: {
      kind: 'Shift',
      state: 10,
    },
    NULL: {
      kind: 'Shift',
      state: 28,
    },
    NUMBER: {
      kind: 'Shift',
      state: 30,
    },
    OPENING_BRACKET: {
      kind: 'Shift',
      state: 23,
    },
    TRUE: {
      kind: 'Shift',
      state: 26,
    },
  },
  {
    FALSE: {
      kind: 'Shift',
      state: 25,
    },
    IDENTIFIER: {
      kind: 'Shift',
      state: 10,
    },
    NULL: {
      kind: 'Shift',
      state: 28,
    },
    NUMBER: {
      kind: 'Shift',
      state: 30,
    },
    OPENING_BRACKET: {
      kind: 'Shift',
      state: 23,
    },
    TRUE: {
      kind: 'Shift',
      state: 26,
    },
  },
  {
    CLOSING_BRACE: {
      kind: 'Shift',
      state: 32,
    },
    IDENTIFIER: {
      kind: 'Shift',
      state: 10,
    },
  },
  {
    ['$']: {
      kind: 'Reduce',
      rule: 21,
    },
  },
  {
    ['$']: {
      kind: 'Reduce',
      rule: 22,
    },
  },
  {
    SEMICOLON: {
      kind: 'Shift',
      state: 37,
    },
  },
  {
    SEMICOLON: {
      kind: 'Reduce',
      rule: 7,
    },
    CLOSING_BRACKET: {
      kind: 'Reduce',
      rule: 7,
    },
    COMMA: {
      kind: 'Reduce',
      rule: 7,
    },
  },
  {
    SEMICOLON: {
      kind: 'Reduce',
      rule: 8,
    },
    CLOSING_BRACKET: {
      kind: 'Reduce',
      rule: 8,
    },
    COMMA: {
      kind: 'Reduce',
      rule: 8,
    },
  },
  {
    SEMICOLON: {
      kind: 'Reduce',
      rule: 9,
    },
    CLOSING_BRACKET: {
      kind: 'Reduce',
      rule: 9,
    },
    COMMA: {
      kind: 'Reduce',
      rule: 9,
    },
  },
  {
    CLOSING_BRACKET: {
      kind: 'Shift',
      state: 38,
    },
    FALSE: {
      kind: 'Shift',
      state: 25,
    },
    IDENTIFIER: {
      kind: 'Shift',
      state: 10,
    },
    NULL: {
      kind: 'Shift',
      state: 28,
    },
    NUMBER: {
      kind: 'Shift',
      state: 30,
    },
    OPENING_BRACKET: {
      kind: 'Shift',
      state: 23,
    },
    TRUE: {
      kind: 'Shift',
      state: 26,
    },
  },
  {
    SEMICOLON: {
      kind: 'Reduce',
      rule: 10,
    },
    CLOSING_BRACKET: {
      kind: 'Reduce',
      rule: 10,
    },
    COMMA: {
      kind: 'Reduce',
      rule: 10,
    },
  },
  {
    SEMICOLON: {
      kind: 'Reduce',
      rule: 17,
    },
    CLOSING_BRACKET: {
      kind: 'Reduce',
      rule: 17,
    },
    COMMA: {
      kind: 'Reduce',
      rule: 17,
    },
  },
  {
    SEMICOLON: {
      kind: 'Reduce',
      rule: 18,
    },
    CLOSING_BRACKET: {
      kind: 'Reduce',
      rule: 18,
    },
    COMMA: {
      kind: 'Reduce',
      rule: 18,
    },
  },
  {
    SEMICOLON: {
      kind: 'Reduce',
      rule: 11,
    },
    CLOSING_BRACKET: {
      kind: 'Reduce',
      rule: 11,
    },
    COMMA: {
      kind: 'Reduce',
      rule: 11,
    },
  },
  {
    SEMICOLON: {
      kind: 'Reduce',
      rule: 19,
    },
    CLOSING_BRACKET: {
      kind: 'Reduce',
      rule: 19,
    },
    COMMA: {
      kind: 'Reduce',
      rule: 19,
    },
  },
  {
    SEMICOLON: {
      kind: 'Reduce',
      rule: 12,
    },
    CLOSING_BRACKET: {
      kind: 'Reduce',
      rule: 12,
    },
    COMMA: {
      kind: 'Reduce',
      rule: 12,
    },
  },
  {
    SEMICOLON: {
      kind: 'Reduce',
      rule: 20,
    },
    CLOSING_BRACKET: {
      kind: 'Reduce',
      rule: 20,
    },
    COMMA: {
      kind: 'Reduce',
      rule: 20,
    },
  },
  {
    SEMICOLON: {
      kind: 'Shift',
      state: 41,
    },
  },
  {
    ['$']: {
      kind: 'Reduce',
      rule: 23,
    },
  },
  {
    CLOSING_BRACE: {
      kind: 'Shift',
      state: 42,
    },
    IDENTIFIER: {
      kind: 'Shift',
      state: 10,
    },
  },
  {
    CLOSING_BRACE: {
      kind: 'Reduce',
      rule: 25,
    },
    IDENTIFIER: {
      kind: 'Reduce',
      rule: 25,
    },
  },
  {
    CLOSING_BRACE: {
      kind: 'Reduce',
      rule: 27,
    },
    IDENTIFIER: {
      kind: 'Reduce',
      rule: 27,
    },
  },
  {
    COLON: {
      kind: 'Shift',
      state: 44,
    },
  },
  {
    ['$']: {
      kind: 'Reduce',
      rule: 4,
    },
  },
  {
    SEMICOLON: {
      kind: 'Reduce',
      rule: 13,
    },
    CLOSING_BRACKET: {
      kind: 'Reduce',
      rule: 13,
    },
    COMMA: {
      kind: 'Reduce',
      rule: 13,
    },
  },
  {
    CLOSING_BRACKET: {
      kind: 'Shift',
      state: 45,
    },
    COMMA: {
      kind: 'Shift',
      state: 46,
    },
  },
  {
    CLOSING_BRACKET: {
      kind: 'Reduce',
      rule: 15,
    },
    COMMA: {
      kind: 'Reduce',
      rule: 15,
    },
  },
  {
    ['$']: {
      kind: 'Reduce',
      rule: 5,
    },
  },
  {
    ['$']: {
      kind: 'Reduce',
      rule: 24,
    },
  },
  {
    CLOSING_BRACE: {
      kind: 'Reduce',
      rule: 26,
    },
    IDENTIFIER: {
      kind: 'Reduce',
      rule: 26,
    },
  },
  {
    IDENTIFIER: {
      kind: 'Shift',
      state: 10,
    },
  },
  {
    SEMICOLON: {
      kind: 'Reduce',
      rule: 14,
    },
    CLOSING_BRACKET: {
      kind: 'Reduce',
      rule: 14,
    },
    COMMA: {
      kind: 'Reduce',
      rule: 14,
    },
  },
  {
    FALSE: {
      kind: 'Shift',
      state: 25,
    },
    IDENTIFIER: {
      kind: 'Shift',
      state: 10,
    },
    NULL: {
      kind: 'Shift',
      state: 28,
    },
    NUMBER: {
      kind: 'Shift',
      state: 30,
    },
    OPENING_BRACKET: {
      kind: 'Shift',
      state: 23,
    },
    TRUE: {
      kind: 'Shift',
      state: 26,
    },
  },
  {
    SEMICOLON: {
      kind: 'Shift',
      state: 49,
    },
  },
  {
    CLOSING_BRACKET: {
      kind: 'Reduce',
      rule: 16,
    },
    COMMA: {
      kind: 'Reduce',
      rule: 16,
    },
  },
  {
    CLOSING_BRACE: {
      kind: 'Reduce',
      rule: 28,
    },
    IDENTIFIER: {
      kind: 'Reduce',
      rule: 28,
    },
  },
];
const gotos = [
  {
    AssignmentStatement: 2,
    ClassDeclaration: 5,
    ExportDefaultDeclaration: 7,
    Statement: 1,
  },
  {},
  {},
  {
    Identifier: 9,
  },
  {
    Identifier: 11,
  },
  {},
  {
    Identifier: 12,
  },
  {},
  {},
  {},
  {},
  {},
  {},
  {
    ClassDeclaration: 18,
    Declaration: 17,
  },
  {
    ArrayValue: 22,
    BooleanValue: 24,
    Expression: 19,
    Identifier: 20,
    NullValue: 27,
    NumberValue: 29,
    PrimitiveValue: 21,
  },
  {
    ArrayValue: 22,
    BooleanValue: 24,
    Expression: 31,
    Identifier: 20,
    NullValue: 27,
    NumberValue: 29,
    PrimitiveValue: 21,
  },
  {
    ClassBodyList: 33,
    ClassBodyListItem: 34,
    Identifier: 36,
    PropertyDeclaration: 35,
  },
  {},
  {},
  {},
  {},
  {},
  {},
  {
    ArrayItemList: 39,
    ArrayValue: 22,
    BooleanValue: 24,
    Expression: 40,
    Identifier: 20,
    NullValue: 27,
    NumberValue: 29,
    PrimitiveValue: 21,
  },
  {},
  {},
  {},
  {},
  {},
  {},
  {},
  {},
  {},
  {
    ClassBodyListItem: 43,
    Identifier: 36,
    PropertyDeclaration: 35,
  },
  {},
  {},
  {},
  {},
  {},
  {},
  {},
  {},
  {},
  {},
  {
    Identifier: 47,
  },
  {},
  {
    ArrayValue: 22,
    BooleanValue: 24,
    Expression: 48,
    Identifier: 20,
    NullValue: 27,
    NumberValue: 29,
    PrimitiveValue: 21,
  },
  {},
  {},
  {},
];
const rules = [
  {
    production: "Statement'",
    pop: 1,
    action: () => {} /* dummy placeholder */,
  },
  {
    production: 'Statement',
    pop: 1,
    action: r1,
  },
  {
    production: 'Statement',
    pop: 1,
    action: r2,
  },
  {
    production: 'Statement',
    pop: 1,
    action: r3,
  },
  {
    production: 'AssignmentStatement',
    pop: 5,
    action: r4,
  },
  {
    production: 'AssignmentStatement',
    pop: 5,
    action: r5,
  },
  {
    production: 'Identifier',
    pop: 1,
    action: r6,
  },
  {
    production: 'Expression',
    pop: 1,
    action: r7,
  },
  {
    production: 'Expression',
    pop: 1,
    action: r8,
  },
  {
    production: 'PrimitiveValue',
    pop: 1,
    action: r9,
  },
  {
    production: 'PrimitiveValue',
    pop: 1,
    action: r10,
  },
  {
    production: 'PrimitiveValue',
    pop: 1,
    action: r11,
  },
  {
    production: 'PrimitiveValue',
    pop: 1,
    action: r12,
  },
  {
    production: 'ArrayValue',
    pop: 2,
    action: r13,
  },
  {
    production: 'ArrayValue',
    pop: 3,
    action: r14,
  },
  {
    production: 'ArrayItemList',
    pop: 1,
    action: r15,
  },
  {
    production: 'ArrayItemList',
    pop: 3,
    action: r16,
  },
  {
    production: 'BooleanValue',
    pop: 1,
    action: r17,
  },
  {
    production: 'BooleanValue',
    pop: 1,
    action: r18,
  },
  {
    production: 'NullValue',
    pop: 1,
    action: r19,
  },
  {
    production: 'NumberValue',
    pop: 1,
    action: r20,
  },
  {
    production: 'ExportDefaultDeclaration',
    pop: 3,
    action: r21,
  },
  {
    production: 'Declaration',
    pop: 1,
    action: r22,
  },
  {
    production: 'ClassDeclaration',
    pop: 4,
    action: r23,
  },
  {
    production: 'ClassDeclaration',
    pop: 5,
    action: r24,
  },
  {
    production: 'ClassBodyList',
    pop: 1,
    action: r25,
  },
  {
    production: 'ClassBodyList',
    pop: 2,
    action: r26,
  },
  {
    production: 'ClassBodyListItem',
    pop: 1,
    action: r27,
  },
  {
    production: 'PropertyDeclaration',
    pop: 4,
    action: r28,
  },
];
const EOF = new Token('$', -1, -1, '');
export default function parseStatement(input) {
  const stack = [[null, 0]];

  const lexer = new Lexer(input);

  let token = lexer.next() || EOF;

  while (true) {
    const [, current] = stack[stack.length - 1];
    const action = actions[current][token.name];

    if (!action) {
      throw new Error('syntax error');
    }
    if (action.kind === 'Accept') {
      // Expect initial state + accept state.
      const [tree] = stack[1];
      return tree;
    } else if (action.kind === 'Shift') {
      stack.push([token, action.state]);
      token = lexer.next() || EOF;
    } else if (action.kind === 'Reduce') {
      // TODO: instead of pop, set length?
      const {production, pop, action: code} = rules[action.rule];
      const popped: Array<P | Token | null> = [];
      for (let i = 0; i < pop; i++) {
        const [node] = stack.pop()!;
        popped[pop - i - 1] = node;
      }
      const [, next] = stack[stack.length - 1];
      const target = gotos[next][production];
      if (code) {
        stack.push([code(...popped), target]);
      } else {
        throw new Error(
          'to use static parser must provide semantic action for every production',
        );
      }
    }
  }
}
