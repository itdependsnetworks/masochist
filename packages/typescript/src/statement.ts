import {
  getAugmentedGrammar,
  getItemSets,
  getParseTable,
  itemSetsToTransitionTable,
  parseDSL,
} from '@masochist/parser';
import Bun from 'bun';
import path from 'path';

const grammarDeclaration = await Bun.file(
  path.join(import.meta.dir, '..', 'src', 'statement.grammar'),
).text();

export const unaugmentedGrammar = parseDSL(grammarDeclaration);
export const grammar = getAugmentedGrammar(unaugmentedGrammar);
export const itemSets = getItemSets(unaugmentedGrammar);
export const transitionTable = itemSetsToTransitionTable(
  itemSets,
  unaugmentedGrammar,
);
export const table = getParseTable(
  itemSets,
  transitionTable,
  unaugmentedGrammar,
);
