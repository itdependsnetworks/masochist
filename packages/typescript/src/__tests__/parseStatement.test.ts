import {getParser} from '@masochist/parser/src/internal';
import {describe, expect, it} from 'bun:test';

import lexer from '../lexer';
import parseStatement from '../parseStatement';
import {grammar, table} from '../statement';

describe('parseStatement()', async () => {
  // We run these tests against a fresh copy of the parser, derived at runtime
  // from "statement.grammar", but also against a copy of the parser as written
  // to "parseStatement.ts". This allows us to do hacky things like inject
  // debug statements into the parser on disk for trouble-shooting purposes, if
  // needed.
  describe.each([
    ['parser derived from grammar', await getParser(grammar, table, lexer)],
    ['parser persisted to disk', parseStatement as any /* TODO: fix type */],
  ])(
    '%s',
    (_description, parseStatement: Awaited<ReturnType<typeof getParser>>) => {
      it('parses a call', () => {
        const input = 'click();';
        expect(parseStatement(input)).toEqual({
          kind: 'ExpressionStatement',
          expression: {
            kind: 'CallExpression',
            arguments: [],
            callee: {
              kind: 'Identifier',
              name: 'click',
            },
          },
        });
      });

      it('parses a `new` expression', () => {
        const input = 'const lexer = new Lexer(input);';
        expect(parseStatement(input)).toEqual({
          kind: 'AssignmentStatement',
          binding: 'const',
          lhs: {
            kind: 'Identifier',
            name: 'lexer',
          },
          rhs: {
            kind: 'NewExpression',
            object: {
              kind: 'Identifier',
              name: 'Lexer',
            },
            arguments: [
              {
                kind: 'Identifier',
                name: 'input',
              },
            ],
          },
        });
      });

      it('parses a const boolean assignment statement', () => {
        const input = 'const isFoo = true;';
        expect(parseStatement(input)).toMatchSnapshot();
      });

      it('parses a let boolean assignment statement', () => {
        const input = 'let isFoo = false;';
        expect(parseStatement(input)).toMatchSnapshot();
      });

      it('parses a const number assignment statement', () => {
        const input = 'const isFoo = 1;';
        expect(parseStatement(input)).toMatchSnapshot();
      });

      it('parses a let number assignment statement', () => {
        const input = 'let isFoo = 2;';
        expect(parseStatement(input)).toMatchSnapshot();
      });

      it('parses a const nested array assignment', () => {
        const input = 'const stack = [[null, 0]];';
        expect(parseStatement(input)).toMatchSnapshot();
      });

      it('parses an empty class declaration', () => {
        const input = 'class Foo {}';
        expect(parseStatement(input)).toEqual({
          kind: 'ClassDeclaration',
          id: 'Foo',
          body: [],
        });
      });

      it('parses a (default) export class declaration', () => {
        const input = `
          export default class Token {
            name: string;
            start: number;
            end: number;
            source: string;

            constructor(name: string, start: number, end: number, source: string) {
              // No validation, for speed; we trust the generated lexer to be flawless.
              this.name = name;
              this.start = start;
              this.end = end;
              this.source = source;
            }

            // Make this a getter. Add rest of real contents to it.
            contents() {
              const value = this.source.slice(this.start, this.end);
              return value;
            }
          }
        `;
        expect(parseStatement(input)).toMatchSnapshot();
        // TODO: add rest of Token.ts example here ^^^
      });

      it('implements right-associativity for chained assignment expressions', () => {
        // Single item in "chain".
        expect(parseStatement('a = b;')).toEqual({
          binding: null,
          kind: 'AssignmentStatement',
          lhs: {
            kind: 'Identifier',
            name: 'a',
          },
          rhs: {
            kind: 'Identifier',
            name: 'b',
          },
        });

        // Two items in "chain".
        expect(parseStatement('a = b = c;')).toEqual({
          binding: null,
          kind: 'AssignmentStatement',
          lhs: {
            kind: 'Identifier',
            name: 'a',
          },
          rhs: {
            kind: 'BinaryExpression',
            lhs: {
              kind: 'Identifier',
              name: 'b',
            },
            operator: '=',
            rhs: {
              kind: 'Identifier',
              name: 'c',
            },
          },
        });

        // Six items in "chain", using another way of showing right-associativity.
        expect(parseStatement('a = b = c = d = e = f;')).toEqual(
          parseStatement('a = (b = (c = (d = (e = f))));'),
        );
      });

      it('implements left-associativity for chained additions', () => {
        // Single item in "chain".
        expect(parseStatement('a + b;')).toEqual({
          kind: 'ExpressionStatement',
          expression: {
            kind: 'BinaryExpression',
            lhs: {
              kind: 'Identifier',
              name: 'a',
            },
            operator: '+',
            rhs: {
              kind: 'Identifier',
              name: 'b',
            },
          },
        });

        // Two items in "chain".
        expect(parseStatement('a + b + c;')).toEqual({
          kind: 'ExpressionStatement',
          expression: {
            kind: 'BinaryExpression',
            lhs: {
              kind: 'BinaryExpression',
              lhs: {
                kind: 'Identifier',
                name: 'a',
              },
              operator: '+',
              rhs: {
                kind: 'Identifier',
                name: 'b',
              },
            },
            operator: '+',
            rhs: {
              kind: 'Identifier',
              name: 'c',
            },
          },
        });

        // Six items in "chain", using another way of showing left-associativity.
        expect(parseStatement('a + b + c + d + e + f;')).toEqual(
          parseStatement('((((a + b) + c) + d) + e) + f;'),
        );
      });

      it('implements left-associativity for chained subtractions', () => {
        // Single item in "chain".
        expect(parseStatement('a - b;')).toEqual({
          kind: 'ExpressionStatement',
          expression: {
            kind: 'BinaryExpression',
            lhs: {
              kind: 'Identifier',
              name: 'a',
            },
            operator: '-',
            rhs: {
              kind: 'Identifier',
              name: 'b',
            },
          },
        });

        // Two items in "chain".
        expect(parseStatement('a - b - c;')).toEqual({
          kind: 'ExpressionStatement',
          expression: {
            kind: 'BinaryExpression',
            lhs: {
              kind: 'BinaryExpression',
              lhs: {
                kind: 'Identifier',
                name: 'a',
              },
              operator: '-',
              rhs: {
                kind: 'Identifier',
                name: 'b',
              },
            },
            operator: '-',
            rhs: {
              kind: 'Identifier',
              name: 'c',
            },
          },
        });

        // Six items in "chain", using another way of showing left-associativity.
        expect(parseStatement('a - b - c - d - e - f;')).toEqual(
          parseStatement('((((a - b) - c) - d) - e) - f;'),
        );
      });

      it('treats plus and minus operators as having the same precedence', () => {
        expect(parseStatement('a - b - c + d + e - f + g;')).toEqual(
          parseStatement('(((((a - b) - c) + d) + e) - f) + g;'),
        );
      });

      it('implements left-associativity for chained member expressions', () => {
        const input = 'const a = b.c.d;';
        const parsed = parseStatement(input);

        expect(parsed).toEqual({
          binding: 'const',
          kind: 'AssignmentStatement',
          lhs: {
            kind: 'Identifier',
            name: 'a',
          },
          rhs: {
            kind: 'MemberExpression',
            object: {
              kind: 'MemberExpression',
              object: {
                kind: 'Identifier',
                name: 'b',
              },
              property: {
                kind: 'Identifier',
                name: 'c',
              },
            },
            property: {
              kind: 'Identifier',
              name: 'd',
            },
          },
        });

        // Showing that we can go deeper:
        expect(parseStatement(`const a = b.c.d.e.f.g;`)).toEqual(
          parseStatement(`const a = ((((b.c).d).e).f).g;`),
        );
      });

      it('parses a chained member expression as a call expression callee', () => {
        const input = 'const a = b.c.d();';
        const parsed = parseStatement(input);

        expect(parsed).toEqual({
          binding: 'const',
          kind: 'AssignmentStatement',
          lhs: {
            kind: 'Identifier',
            name: 'a',
          },
          rhs: {
            kind: 'CallExpression',
            arguments: [],
            callee: {
              kind: 'MemberExpression',
              object: {
                kind: 'MemberExpression',
                object: {
                  kind: 'Identifier',
                  name: 'b',
                },
                property: {
                  kind: 'Identifier',
                  name: 'c',
                },
              },
              property: {
                kind: 'Identifier',
                name: 'd',
              },
            },
          },
        });
      });
    },
  );
});
