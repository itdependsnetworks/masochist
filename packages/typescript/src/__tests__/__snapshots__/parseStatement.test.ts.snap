// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`parseStatement() parses a const boolean assignment statement 1`] = `
{
  "binding": "const",
  "kind": "AssignmentStatement",
  "lhs": {
    "kind": "Identifier",
    "name": "isFoo",
  },
  "rhs": {
    "kind": "BooleanValue",
    "value": true,
  },
}
`;

exports[`parseStatement() parses a let boolean assignment statement 1`] = `
{
  "binding": "let",
  "kind": "AssignmentStatement",
  "lhs": {
    "kind": "Identifier",
    "name": "isFoo",
  },
  "rhs": {
    "kind": "BooleanValue",
    "value": false,
  },
}
`;

exports[`parseStatement() parses a const number assignment statement 1`] = `
{
  "binding": "const",
  "kind": "AssignmentStatement",
  "lhs": {
    "kind": "Identifier",
    "name": "isFoo",
  },
  "rhs": {
    "base": 10,
    "kind": "NumberValue",
    "value": 1,
  },
}
`;

exports[`parseStatement() parses a let number assignment statement 1`] = `
{
  "binding": "let",
  "kind": "AssignmentStatement",
  "lhs": {
    "kind": "Identifier",
    "name": "isFoo",
  },
  "rhs": {
    "base": 10,
    "kind": "NumberValue",
    "value": 2,
  },
}
`;

exports[`parseStatement() parses a const nested array assignment 1`] = `
{
  "binding": "const",
  "kind": "AssignmentStatement",
  "lhs": {
    "kind": "Identifier",
    "name": "stack",
  },
  "rhs": {
    "items": [
      {
        "items": [
          {
            "kind": "NullValue",
          },
          {
            "base": 10,
            "kind": "NumberValue",
            "value": 0,
          },
        ],
        "kind": "ArrayValue",
      },
    ],
    "kind": "ArrayValue",
  },
}
`;

exports[`parseStatement() parses a (default) export class declaration 1`] = `
{
  "declaration": {
    "body": [
      {
        "kind": "PropertyDeclaration",
        "name": "name",
        "type": "string",
      },
      {
        "kind": "PropertyDeclaration",
        "name": "start",
        "type": "number",
      },
      {
        "kind": "PropertyDeclaration",
        "name": "end",
        "type": "number",
      },
      {
        "kind": "PropertyDeclaration",
        "name": "source",
        "type": "string",
      },
    ],
    "id": "Token",
    "kind": "ClassDeclaration",
  },
  "kind": "ExportDefaultDeclaration",
}
`;
